ring_theory/hilbert_basis.lean

```diff
@@ -0,0 +1,238 @@
+leading coefficients of polynomials in `I` with degree ≤ `n`. -/
+def leading_coeff_nth (n : ℕ) : ideal R :=
+(I.degree_le n).map $ lcoeff R n
+
+theorem mem_leading_coeff_nth (n : ℕ) (x) :
+  x ∈ I.leading_coeff_nth n ↔ ∃ p ∈ I, degree p ≤ n ∧ leading_coeff p = x :=
+begin
+  simp only [leading_coeff_nth, degree_le, submodule.mem_map, lcoeff_apply, submodule.mem_inf, mem_degree_le],
+  split,
+  { rintro ⟨p, ⟨hpdeg, hpI⟩, rfl⟩,
+    cases lt_or_eq_of_le hpdeg with hpdeg hpdeg,
+    { refine ⟨0, I.zero_mem, lattice.bot_le, _⟩,
+      rw [leading_coeff_zero, eq_comm],
+      exact coeff_eq_zero_of_degree_lt hpdeg },
+    { refine ⟨p, hpI, le_of_eq hpdeg, _⟩,
+      rw [leading_coeff, nat_degree, hpdeg], refl } },
+  { rintro ⟨p, hpI, hpdeg, rfl⟩,
+    have : nat_degree p + (n - nat_degree p) = n,
+    { exact nat.add_sub_cancel' (nat_degree_le_of_degree_le hpdeg) },
+    refine ⟨p * X ^ (n - nat_degree p), ⟨_, I.mul_mem_right hpI⟩, _⟩,
+    { apply le_trans (degree_mul_le _ _) _,
+      apply le_trans (add_le_add' (degree_le_nat_degree) (degree_X_pow_le _)) _,
+      rw [← with_bot.coe_add, this],
+      exact le_refl _ },
+    { rw [leading_coeff, ← coeff_mul_X_pow p (n - nat_degree p), this] } }
+end
+
+theorem mem_leading_coeff_nth_zero (x) :
+  x ∈ I.leading_coeff_nth 0 ↔ C x ∈ I :=
+(mem_leading_coeff_nth _ _ _).trans
+⟨λ ⟨p, hpI, hpdeg, hpx⟩, by rwa [← hpx, leading_coeff,
+  nat.eq_zero_of_le_zero (nat_degree_le_of_degree_le hpdeg),
+  ← eq_C_of_degree_le_zero hpdeg],
+λ hx, ⟨C x, hx, degree_C_le, leading_coeff_C x⟩⟩
+
+theorem leading_coeff_nth_mono {m n : ℕ} (H : m ≤ n) :
+  I.leading_coeff_nth m ≤ I.leading_coeff_nth n :=
+begin
+  intros r hr,
+  simp only [submodule.mem_coe, mem_leading_coeff_nth] at hr ⊢,
+  rcases hr with ⟨p, hpI, hpdeg, rfl⟩,
+  refine ⟨p * X ^ (n - m), I.mul_mem_right hpI, _, leading_coeff_mul_X_pow⟩,
+  refine le_trans (degree_mul_le _ _) _,
+  refine le_trans (add_le_add' hpdeg (degree_X_pow_le _)) _,
+  rw [← with_bot.coe_add, nat.add_sub_cancel' H],
+  exact le_refl _
+end
+
+/-- Given an ideal `I` in `R[X]`, make the ideal in `R` of the
+leading coefficients in `I`. -/
+def leading_coeff : ideal R :=
+⨆ n : ℕ, I.leading_coeff_nth n
+
+theorem mem_leading_coeff (x) :
+  x ∈ I.leading_coeff ↔ ∃ p ∈ I, polynomial.leading_coeff p = x :=
+begin
+  rw [leading_coeff, submodule.mem_supr_of_directed],
+  simp only [mem_leading_coeff_nth],
+  { split, { rintro ⟨i, p, hpI, hpdeg, rfl⟩, exact ⟨p, hpI, rfl⟩ },
+    rintro ⟨p, hpI, rfl⟩, exact ⟨nat_degree p, p, hpI, degree_le_nat_degree, rfl⟩ },
+  { exact ⟨0⟩ },
+  intros i j, exact ⟨i + j, I.leading_coeff_nth_mono (nat.le_add_right _ _),
+    I.leading_coeff_nth_mono (nat.le_add_left _ _)⟩
+end
+
+theorem is_fg_degree_le (hnr : is_noetherian_ring R) (n : ℕ) :
+  submodule.fg (I.degree_le n) :=
+is_noetherian_submodule_left.1 (is_noetherian_of_fg_of_noetherian _ hnr
+  ⟨_, degree_le_eq_span_X_pow.symm⟩) _
+
+end ideal
+
+/-- Hilbert basis theorem. -/
```

Comment:
I guess, this should move to `ring_theory.noetherian`

Reply:
So the ideal stuff above it goes to limbo?